### ISSUE TITLE
After adding a new generation of patterns (acls), the CLI command 'get acl' does not give the correct result

### ISSUE THREAD

**Issue by zaga00** [haproxy/haproxy#1802]
### Detailed Description of the Problem

After creating a new generation of patterns (acls) and adding new ones (and before switching the active generation to a new one, i.e. commit), the CLI command 'get acl' returns an incorrect result.  After commit, the CLI command works correctly.

The bug exists on all versions of haproxy newer than 2.4 (including this one).

```% echo "show acl" | socat - "UNIX-CONNECT:/tmp/haproxy.sock"
# id (file) description
0 (data.txt) pattern loaded from file 'data.txt' used by acl at file 'haproxy.cfg' line 14. curr_ver=0 next_ver=0 entry_cnt=1
1 () acl 'var' file 'haproxy.cfg' line 14. curr_ver=0 next_ver=0 entry_cnt=0
2 () acl 'var' file 'httpclient' line 0. curr_ver=0 next_ver=0 entry_cnt=1
3 () acl 'var' file 'httpclient' line 0. curr_ver=0 next_ver=0 entry_cnt=1


% echo "get acl #0 66.56.56.135" | socat - "UNIX-CONNECT:/tmp/haproxy.sock"
type=ip, case=sensitive, match=no


% echo "prepare acl #0" | socat - "UNIX-CONNECT:/tmp/haproxy.sock"
New version created: 1


% echo "add acl @1 #0 66.56.0.0/18" | socat - "UNIX-CONNECT:/tmp/haproxy.sock"


% echo "show acl" | socat - "UNIX-CONNECT:/tmp/haproxy.sock"
# id (file) description
0 (data.txt) pattern loaded from file 'data.txt' used by acl at file 'haproxy.cfg' line 14. curr_ver=0 next_ver=1 entry_cnt=2
1 () acl 'var' file 'haproxy.cfg' line 14. curr_ver=0 next_ver=0 entry_cnt=0
2 () acl 'var' file 'httpclient' line 0. curr_ver=0 next_ver=0 entry_cnt=1
3 () acl 'var' file 'httpclient' line 0. curr_ver=0 next_ver=0 entry_cnt=1

--- INVALID RESULT ---
% echo "get acl #0 66.56.56.135" | socat - "UNIX-CONNECT:/tmp/haproxy.sock"
type=ip, case=sensitive, match=yes, idx=tree, pattern="66.59.61.0/24"


% echo "commit acl @1 #0" | socat - "UNIX-CONNECT:/tmp/haproxy.sock"


% echo "show acl" | socat - "UNIX-CONNECT:/tmp/haproxy.sock"
# id (file) description
0 (data.txt) pattern loaded from file 'data.txt' used by acl at file 'haproxy.cfg' line 14. curr_ver=1 next_ver=1 entry_cnt=1
1 () acl 'var' file 'haproxy.cfg' line 14. curr_ver=0 next_ver=0 entry_cnt=0
2 () acl 'var' file 'httpclient' line 0. curr_ver=0 next_ver=0 entry_cnt=1
3 () acl 'var' file 'httpclient' line 0. curr_ver=0 next_ver=0 entry_cnt=1


% echo "get acl #0 66.56.56.135" | socat - "UNIX-CONNECT:/tmp/haproxy.sock"
type=ip, case=sensitive, match=yes, idx=tree, pattern="66.56.0.0/18"
```

### Expected Behavior

The pattern search should only be performed on the currently active generation.

The only difference is that after adding the acl (and before the commit), the cli command 'get acl ..' should give the correct result (which is `type=ip, case=sensitive, match=no`):

```% echo "show acl" ..
% echo "get acl #0 66.56.56.135" ..
% echo "prepare acl #0" ..
% echo "add acl @1 #0 66.56.0.0/18" ..
% echo "show acl" ..


% echo "get acl #0 66.56.56.135" | socat - "UNIX-CONNECT:/tmp/haproxy.sock"
type=ip, case=sensitive, match=no


% echo "commit acl @1 #0" ..
% echo "show acl" ..


% echo "get acl #0 66.56.56.135" | socat - "UNIX-CONNECT:/tmp/haproxy.sock"
type=ip, case=sensitive, match=yes, idx=tree, pattern="66.56.0.0/18"
```

### Steps to Reproduce the Behavior

1. start haproxy: `./haproxy -f haproxy.cfg`
2. in another shell, run the commands as in the problem description

### Do you have any idea what may have caused this?

The problem is in the cli_io_handler_map_lookup() function, i.e. probably deeper in the source regarding pattern matching.  It looks like the pattern search function searches all generations and not just the one that is currently active.

### Do you have an idea how to solve the issue?

The pattern search should only be performed on the currently active generation.

### What is your configuration?

```haproxy
global
    stats socket /tmp/haproxy.sock mode 666 level admin
    log stdout daemon

defaults
    log global
    mode http
    timeout connect 5000
    timeout client  50000
    timeout server  50000

frontend f1
    bind *:10080
    acl is_data var(txn.qsrc) -m ip -f data.txt
    http-request set-var(txn.qsrc) urlp(src)
    http-request set-var(txn.data) int(1) if is_data
```

### Output of `haproxy -vv`

```plain
% ./haproxy -vv
HAProxy version 2.7-dev2-53bfac-19 2022/07/22 - https://haproxy.org/
Status: development branch - not safe for use in production.
Known bugs: https://github.com/haproxy/haproxy/issues?q=is:issue+is:open
Running on: Linux 5.16.0-0.bpo.4-amd64 #1 SMP PREEMPT Debian 5.16.12-1~bpo11+1 (2022-03-08) x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = 
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT -PCRE2 -PCRE2_JIT +POLL +THREAD +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -ENGINE +GETADDRINFO -OPENSSL -LUA +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC -PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=16).
Built with network namespace support.
Support for malloc_trim() is enabled.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built without PCRE or PCRE2 support (using libc's regex instead)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 10.2.1 20210110

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : none

Available filters :
        [BWLIM] bwlim-in
        [BWLIM] bwlim-out
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app

% uname -a
Linux tirnanog 5.16.0-0.bpo.4-amd64 #1 SMP PREEMPT Debian 5.16.12-1~bpo11+1 (2022-03-08) x86_64 GNU/Linux
```

### Additional Information

data.txt file with initial acls:

[data.txt](https://github.com/haproxy/haproxy/files/9179702/data.txt)

**Comment by wtarreau** [comment]
For me you're observing the normal and intended behavior. The principle is that "get acl xxxx" will use the same lookup mechanism that the rest of the traffic uses, that applies to the current generation and not any of the next ones currently being uploadedd. Once you commit the new generation it becomes the new current one, and atomically all lookups will apply to this new generation and not the old one. But before the commit that's expected. In short, the "commit" operation does nothing more than "current := next" and start the purge in the background of anything older than current.

**Comment by wtarreau** [comment]
OK forget my reply above, Christopher just showed me that I misread the output. I didn't notice that the returned IP differs on the 3rd byte. Christopher found the cause of the problem, it's due to the `continue` statement I added to find other generations, it's just that it forgets to re-compare :-(  We're on it. Thank you for your report, Miroslav!

**Comment by wtarreau** [comment]
For the record, the faulty commit is c93da6950 ("MEDIUM: pattern: only match patterns that match the current generation")  (introduced in 2020 in 2.4-dev1).

**Comment by wtarreau** [comment]
OK now fixed by commit 51d38a26f, I'm going to backport it. Thank you Miroslav for reporting this mess!

**Comment by zaga00** [comment]
It works ok now, thank you.
